pkg/loadbalancer/experimental/services.go

```diff
@@ -0,0 +1,399 @@
+// implement "resynchronization", for example with K8s when the Watch() call fails and we need
+// to start over with a List().
+func (s *Services) DeleteFrontendsBySource(txn ServiceWriteTxn, source source.Source) error {
+	// Iterating over all as this is a rare operation and it would be costly
+	// to always index by source.
+	iter, _ := s.fes.All(txn)
+	for svc, _, ok := iter.Next(); ok; svc, _, ok = iter.Next() {
+		if svc.Source == source {
+			if err := s.deleteFrontend(txn, svc); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (s *Services) UpsertBackends(txn ServiceWriteTxn, serviceName loadbalancer.ServiceName, source source.Source, bes ...*loadbalancer.Backend) error {
+	if err := s.updateBackends(txn, serviceName, source, bes); err != nil {
+		return err
+	}
+
+	iter := s.fes.List(txn, FrontendNameIndex.Query(serviceName))
+	for fe, _, ok := iter.Next(); ok; fe, _, ok = iter.Next() {
+		fe = fe.Clone()
+		fe.Status = reconciler.StatusPending()
+		fe.BackendRevision = s.bes.Revision(txn)
+		if _, _, err := s.fes.Insert(txn, fe); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func NewServiceNameSet(names ...loadbalancer.ServiceName) container.ImmSet[loadbalancer.ServiceName] {
+	return container.NewImmSetFunc(
+		loadbalancer.ServiceName.Compare,
+		names...,
+	)
+}
+
+func (s *Services) updateBackends(txn ServiceWriteTxn, serviceName loadbalancer.ServiceName, source source.Source, bes []*loadbalancer.Backend) error {
+	for _, bep := range bes {
+		var be Backend
+		if old, _, ok := s.bes.Get(txn, BackendAddrIndex.Query(bep.L3n4Addr)); ok {
+			// Keep the old state, e.g. ID
+			be = *old
+		} else {
+			be.ReferencedBy = NewServiceNameSet(serviceName)
+		}
+		be.Backend = *bep
+		be.Source = source
+		be.ReferencedBy = be.ReferencedBy.Insert(serviceName)
+
+		// TODO: Here we should figure out how to merge the two states. E.g. we might learn
+		// through health checking that the backend should be inactive, or from k8s that the
+		// backend is terminating (which overrides everything).
+		// See Service.updateBackendsCacheLocked.
+		be.ActualState = bep.State
+
+		if _, _, err := s.bes.Insert(txn, &be); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Services) DeleteBackendsBySource(txn ServiceWriteTxn, source source.Source) error {
+	// Iterating over all as this is a rare operation and it would be costly
+	// to always index by source.
+	names := sets.New[loadbalancer.ServiceName]()
+	iter, _ := s.bes.All(txn)
+	for be, _, ok := iter.Next(); ok; be, _, ok = iter.Next() {
+		if be.Source == source {
+			names.Insert(be.ReferencedBy.AsSlice()...)
+			if _, _, err := s.bes.Delete(txn, be); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Bump the backend revision for each of the referenced services to force
+	// reconciliation.
+	revision := s.bes.Revision(txn)
+	for name := range names {
+		fe, _, found := s.fes.Get(txn, FrontendNameIndex.Query(name))
```

Comment:
In line 335 we use the `List` method for this index, and the index has `Unique: false` in `pkg/loadbalancer/experimental/tables.go`, line 253, so I'd suppose we should use `List` instead of `Get` also here.